app/models/ems_refresh/save_inventory.rb

```diff
@@ -205,17 +205,23 @@ def save_guest_devices_inventory(hardware, hashes)
     store_ids_for_new_records(hardware.guest_devices, hashes, [:device_type, :uid_ems])
   end
 
-  def save_disks_inventory(hardware, hashes)
+  def save_disks_inventory(parent, hashes)
     return if hashes.nil?
 
     # Update the associated ids
     hashes.each do |h|
-      h[:storage_id]         = h.fetch_path(:storage, :id)
+      if h[:storage_id].nil?
+        h[:storage_id] = h.fetch_path(:storage, :id)
+      end
       h[:backing_id]         = h.fetch_path(:backing, :id)
       h[:storage_profile_id] = h.fetch_path(:storage_profile, :id)
     end
 
-    save_inventory_multi(hardware.disks, hashes, :use_association, [:controller_type, :location], nil, [:storage, :backing, :storage_profile])
+    filenames = hashes.collect { |h| h[:filename] }
+
+    all_disks = Disk.where(:filename => filenames)
+    save_inventory_multi_child(:disks, all_disks, parent, hashes, [:controller_type, :location, :device_type],
```

Comment:
so all you changes in here and app/models/ems_refresh/save_inventory_helper.rb  should not be necessary. And since this is a shared code with all infra, I am afraid we can;t change it like this.

From the code it seems to me you are looking just for a different call of save_inventory_multi(hardware.disks, hashes, :use_association, [:controller_type, :location], nil, [:storage, :backing, :storage_profile])

Since your key is a filename it should be enough to call
save_inventory_multi(hardware.disks, hashes, :use_association, [:filename], nil, [:storage, :backing, :storage_profile])

And you should condition this only for RHEV provider, somewhere in  
save_hardware_inventory you could check some condition on Vm and pass it in this method, than change the 'unique key' attributes based on this.

We already do something like this here https://github.com/Ladas/manageiq/blob/19f68e729aba475a68fb0a526fb9b2516a73b511/app/models/ems_refresh/save_inventory_network.rb#L189-L189

Instead of checking for type, you could rather set a method on base VM and RHEV Vm, indicating the volume index. Maybe something like 'deterministic_disk_location?', since that seems to be the problem here? You don't get the disks always in the same order, so location can be mapped to different disks?